
# **Problem Statement:**
**Building a new Employee Security System**
You, as the Security analyst, at Stark Industries, have been tasked to build a new contactless employee check-in system. Currently the employees use a physical keycard for entry into the building like shown below. 

You have come up with a new idea that uses the employees smartphone and machine learning to provide a contactless system where when an employee enters the firm’s territory, his or her smartphone connects to the server and transmits data from the employee smartphone sensor data like the accelerometer's data. The server performs the calculations and determines this person as
one of the employees using Gait analysis. Essentially it compares the current pattern of the employee’s gait with the historial pattern and if there is a match, the doors automatically open for the employee to walk in.

To test your idea, you have built a dataset of 30 employees and their daily activities here. Design and develop a system that will perform the gait analysis. 

# **GAIT Analysis**



In [1]:
!unzip "/content/UCI HAR Dataset.zip"

'unzip' is not recognized as an internal or external command,
operable program or batch file.


### Logistic Regression

In [5]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

In [6]:
train_df = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\X_train.txt", delim_whitespace = True,names=list(range(0,561)))

train_df['Subject'] = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\subject_train.txt", header = None, squeeze = True)

train_df["activity"] = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\y_train.txt", header = None, squeeze = True)
# C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train

In [7]:
features = list()
with open(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset/features.txt") as f:
    for line in f:
        features.append(line.split()[1])

In [8]:
features

['tBodyAcc-mean()-X',
 'tBodyAcc-mean()-Y',
 'tBodyAcc-mean()-Z',
 'tBodyAcc-std()-X',
 'tBodyAcc-std()-Y',
 'tBodyAcc-std()-Z',
 'tBodyAcc-mad()-X',
 'tBodyAcc-mad()-Y',
 'tBodyAcc-mad()-Z',
 'tBodyAcc-max()-X',
 'tBodyAcc-max()-Y',
 'tBodyAcc-max()-Z',
 'tBodyAcc-min()-X',
 'tBodyAcc-min()-Y',
 'tBodyAcc-min()-Z',
 'tBodyAcc-sma()',
 'tBodyAcc-energy()-X',
 'tBodyAcc-energy()-Y',
 'tBodyAcc-energy()-Z',
 'tBodyAcc-iqr()-X',
 'tBodyAcc-iqr()-Y',
 'tBodyAcc-iqr()-Z',
 'tBodyAcc-entropy()-X',
 'tBodyAcc-entropy()-Y',
 'tBodyAcc-entropy()-Z',
 'tBodyAcc-arCoeff()-X,1',
 'tBodyAcc-arCoeff()-X,2',
 'tBodyAcc-arCoeff()-X,3',
 'tBodyAcc-arCoeff()-X,4',
 'tBodyAcc-arCoeff()-Y,1',
 'tBodyAcc-arCoeff()-Y,2',
 'tBodyAcc-arCoeff()-Y,3',
 'tBodyAcc-arCoeff()-Y,4',
 'tBodyAcc-arCoeff()-Z,1',
 'tBodyAcc-arCoeff()-Z,2',
 'tBodyAcc-arCoeff()-Z,3',
 'tBodyAcc-arCoeff()-Z,4',
 'tBodyAcc-correlation()-X,Y',
 'tBodyAcc-correlation()-X,Z',
 'tBodyAcc-correlation()-Y,Z',
 'tGravityAcc-mean()-X',
 'tGravityA

In [9]:
test_df = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\test\X_test.txt", delim_whitespace = True,names=list(range(0,561)))

test_df['Subject'] = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\test\subject_test.txt", header = None, squeeze = True)

test_df["activity"] = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\test\y_test.txt", header = None, squeeze = True)

In [10]:
train_df.to_csv('train.csv',index_label=False)

In [11]:
test_df.to_csv('test.csv',index_label=False)

In [12]:
train_df= pd.read_csv('train.csv')

In [13]:
train_df.head()

0         1         2         3         4         5         6  \
0  0.288585 -0.020294 -0.132905 -0.995279 -0.983111 -0.913526 -0.995112   
1  0.278419 -0.016411 -0.123520 -0.998245 -0.975300 -0.960322 -0.998807   
2  0.279653 -0.019467 -0.113462 -0.995380 -0.967187 -0.978944 -0.996520   
3  0.279174 -0.026201 -0.123283 -0.996091 -0.983403 -0.990675 -0.997099   
4  0.276629 -0.016570 -0.115362 -0.998139 -0.980817 -0.990482 -0.998321   

          7         8         9  ...       553       554       555       556  \
0 -0.983185 -0.923527 -0.934724  ... -0.710304 -0.112754  0.030400 -0.464761   
1 -0.974914 -0.957686 -0.943068  ... -0.861499  0.053477 -0.007435 -0.732626   
2 -0.963668 -0.977469 -0.938692  ... -0.760104 -0.118559  0.177899  0.100699   
3 -0.982750 -0.989302 -0.938692  ... -0.482845 -0.036788 -0.012892  0.640011   
4 -0.979672 -0.990441 -0.942469  ... -0.699205  0.123320  0.122542  0.693578   

        557       558       559       560  Subject  activity  
0 -0.018446 -0.841247  0.179941 -0.058627        1         5  
1  0.703511 -0.844788  0.180289 -0.054317        1         5  
2  0.808529 -0.848933  0.180637 -0.049118        1         5  
3 -0.485366 -0.848649  0.181935 -0.047663        1         5  
4 -0.615971 -0.847865  0.185151 -0.043892        1         5  

[5 rows x 563 columns]

In [14]:
test_df= pd.read_csv('test.csv')

In [15]:
test_df.head()

0         1         2         3         4         5         6  \
0  0.257178 -0.023285 -0.014654 -0.938404 -0.920091 -0.667683 -0.952501   
1  0.286027 -0.013163 -0.119083 -0.975415 -0.967458 -0.944958 -0.986799   
2  0.275485 -0.026050 -0.118152 -0.993819 -0.969926 -0.962748 -0.994403   
3  0.270298 -0.032614 -0.117520 -0.994743 -0.973268 -0.967091 -0.995274   
4  0.274833 -0.027848 -0.129527 -0.993852 -0.967445 -0.978295 -0.994111   

          7         8         9  ...       553       554       555       556  \
0 -0.925249 -0.674302 -0.894088  ... -0.705974  0.006462  0.162920 -0.825886   
1 -0.968401 -0.945823 -0.894088  ... -0.594944 -0.083495  0.017500 -0.434375   
2 -0.970735 -0.963483 -0.939260  ... -0.640736 -0.034956  0.202302  0.064103   
3 -0.974471 -0.968897 -0.938610  ... -0.736124 -0.017067  0.154438  0.340134   
4 -0.965953 -0.977346 -0.938610  ... -0.846595 -0.002223 -0.040046  0.736715   

        557       558       559       560  Subject  activity  
0  0.271151 -0.720009  0.276801 -0.057978        2         5  
1  0.920593 -0.698091  0.281343 -0.083898        2         5  
2  0.145068 -0.702771  0.280083 -0.079346        2         5  
3  0.296407 -0.698954  0.284114 -0.077108        2         5  
4 -0.118545 -0.692245  0.290722 -0.073857        2         5  

[5 rows x 563 columns]

In [16]:
from sklearn.model_selection import train_test_split
x=train_df.drop('activity',axis=1)
y=train_df['activity']
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size= 0.30)


In [17]:
X_train

0         1         2         3         4         5         6  \
2578  0.276659 -0.017972 -0.109811 -0.997301 -0.996563 -0.991722 -0.997907   
5281  0.277434 -0.016646 -0.110331 -0.997560 -0.978725 -0.984365 -0.997517   
4714  0.330832 -0.019251 -0.000440 -0.984174 -0.934407 -0.905004 -0.984969   
2750  0.287983 -0.035961 -0.099700 -0.985907 -0.953465 -0.969849 -0.989593   
6081  0.276935 -0.011099 -0.107398 -0.995608 -0.979550 -0.992347 -0.995820   
...        ...       ...       ...       ...       ...       ...       ...   
6123  0.276902 -0.017565 -0.102219 -0.996272 -0.984840 -0.967623 -0.996120   
3254  0.275303 -0.016987 -0.105997 -0.996526 -0.987243 -0.995171 -0.997547   
1189  0.272183 -0.018254 -0.101465 -0.988437 -0.975221 -0.970075 -0.990743   
2571  0.284256 -0.049886 -0.195765 -0.979544 -0.889374 -0.861336 -0.981286   
3848  0.236400 -0.015105 -0.015229 -0.967508 -0.894758 -0.861062 -0.970986   

             7         8         9  ...       552       553       554  \
2578 -0.996629 -0.992091 -0.941745  ... -0.676360 -0.867270  0.481826   
5281 -0.980234 -0.984309 -0.942947  ... -0.450441 -0.809991  0.115075   
4714 -0.941181 -0.904175 -0.918061  ... -0.077340 -0.331037  0.013684   
2750 -0.954323 -0.965256 -0.917344  ...  0.237138  0.009815 -0.031298   
6081 -0.979039 -0.991317 -0.942464  ... -0.678890 -0.898185  0.141837   
...        ...       ...       ...  ...       ...       ...       ...   
6123 -0.982946 -0.963923 -0.942070  ... -0.782953 -0.962545  0.017762   
3254 -0.985770 -0.995856 -0.942428  ... -0.462171 -0.687313  0.539495   
1189 -0.974996 -0.972269 -0.928476  ... -0.202444 -0.480920  0.321764   
2571 -0.875178 -0.837461 -0.927202  ... -0.534703 -0.787254  0.271126   
3848 -0.913319 -0.858411 -0.922744  ...  0.433676  0.207307  0.026719   

           555       556       557       558       559       560  Subject  
2578 -0.483197 -0.292536 -0.304065 -0.476969 -0.229382 -0.136844       15  
5281  0.396187  0.212884  0.045780 -0.839054  0.192084 -0.044347       25  
4714  0.729994 -0.040568  0.378877 -0.648902  0.202737  0.244674       23  
2750 -0.375060 -0.207688  0.375508 -0.853115 -0.004725  0.040763       15  
6081 -0.119057  0.805316 -0.091454 -0.828289  0.213068  0.048697       27  
...        ...       ...       ...       ...       ...       ...      ...  
6123  0.354081  0.133578 -0.208856 -0.879640  0.023261 -0.027463       27  
3254 -0.320092 -0.207466  0.819625 -0.670074  0.218545  0.217948       17  
1189 -0.498381  0.731490  0.791581 -0.834937 -0.019338  0.027106        6  
2571 -0.324476  0.077134 -0.091175 -0.458243 -0.231866 -0.158737       15  
3848 -0.168743  0.222293 -0.336632 -0.668163 -0.071895 -0.143703       19  

[5146 rows x 562 columns]

In [18]:
X_test

0         1         2         3         4         5         6  \
1551  0.224801 -0.030660 -0.126833 -0.194398  0.214787  0.109286 -0.219858   
6753  0.138144 -0.057711 -0.032959  0.119746 -0.307043 -0.242437  0.035522   
5186  0.258364  0.001680 -0.092387 -0.610562 -0.194850 -0.489910 -0.655642   
4623  0.337784 -0.011351 -0.122571 -0.966281 -0.978032 -0.972902 -0.966471   
4260  0.209164 -0.020974 -0.077783 -0.927778 -0.883464 -0.901811 -0.927994   
...        ...       ...       ...       ...       ...       ...       ...   
4321  0.271782 -0.013605 -0.094927 -0.191420  0.073331 -0.048479 -0.243782   
1638  0.280194 -0.018493 -0.104641 -0.998817 -0.986561 -0.979334 -0.999160   
6819  0.275382 -0.019205 -0.108602 -0.996153 -0.974959 -0.992048 -0.996407   
4019  0.279391 -0.015785 -0.106483 -0.997477 -0.992278 -0.992815 -0.997509   
6535  0.143105  0.020477 -0.181751 -0.792843 -0.512246 -0.587244 -0.843663   

             7         8         9  ...       552       553       554  \
1551  0.164404  0.067963  0.292155  ...  0.130098 -0.220469  0.485537   
6753 -0.348022 -0.226344  0.525368  ...  0.221470 -0.140951  0.395421   
5186 -0.187765 -0.481728 -0.438947  ...  0.377739  0.136212  0.198191   
4623 -0.976766 -0.970323 -0.893800  ... -0.272442 -0.631119 -0.020591   
4260 -0.905714 -0.897349 -0.901746  ... -0.396796 -0.792877 -0.099329   
...        ...       ...       ...  ...       ...       ...       ...   
4321  0.055832 -0.036678  0.146847  ...  0.235791 -0.050494  0.369512   
1638 -0.984644 -0.978182 -0.940426  ... -0.786564 -0.931182 -0.240313   
6819 -0.973246 -0.991743 -0.942366  ... -0.552864 -0.849703  0.194412   
4019 -0.991501 -0.994067 -0.941357  ... -0.725897 -0.902829 -0.114875   
6535 -0.472478 -0.602030 -0.803513  ...  0.517458  0.205012  0.222368   

           555       556       557       558       559       560  Subject  
1551  0.697914 -0.747526  0.222649 -0.850565  0.201266  0.010959        7  
6753 -0.502067 -0.921945 -0.822353 -0.887145  0.142678 -0.049743       29  
5186 -0.032256 -0.789118  0.665929 -0.698525  0.177606 -0.182661       25  
4623 -0.233127  0.274191  0.053722  0.481134 -0.499527 -0.498840       22  
4260 -0.140616  0.230059  0.023006  0.342046 -0.370074 -0.598060       21  
...        ...       ...       ...       ...       ...       ...      ...  
4321 -0.403600 -0.460845 -0.271926 -0.578504  0.346753  0.169023       21  
1638  0.021304 -0.150568  0.145345 -0.845233  0.197042  0.060404        8  
6819 -0.763233 -0.797991 -0.186229 -0.900695  0.166062  0.028869       29  
4019  0.316709 -0.425101 -0.440782 -0.738950  0.115121  0.206200       21  
6535 -0.366016 -0.219877  0.050104  0.465025 -0.284737 -0.718141       28  

[2206 rows x 562 columns]

In [19]:
X_train.shape, X_test.shape

((5146, 562), (2206, 562))

In [20]:
y_train.shape, y_test.shape

((5146,), (2206,))

In [21]:
from sklearn.linear_model import LogisticRegression
model= LogisticRegression(max_iter=5000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=5000)

In [22]:
model.score(X_train,y_train)

0.9937815779246016

In [23]:
pred=model.predict(X_test)

In [24]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, pred)

0.9877606527651859

In [25]:
from sklearn.metrics import classification_report
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           1       0.99      1.00      1.00       358
           2       1.00      1.00      1.00       308
           3       1.00      1.00      1.00       280
           4       0.98      0.97      0.97       412
           5       0.97      0.98      0.97       427
           6       1.00      1.00      1.00       421

    accuracy                           0.99      2206
   macro avg       0.99      0.99      0.99      2206
weighted avg       0.99      0.99      0.99      2206



In [26]:
from sklearn.metrics import plot_confusion_matrix, confusion_matrix
confusion_matrix(y_test, pred)

array([[357,   0,   1,   0,   0,   0],
       [  1, 307,   0,   0,   0,   0],
       [  1,   0, 279,   0,   0,   0],
       [  0,   0,   0, 398,  13,   1],
       [  0,   0,   0,  10, 417,   0],
       [  0,   0,   0,   0,   0, 421]], dtype=int64)

In [27]:
from sklearn.metrics import f1_score
f1_score(y_test, pred, average='macro')

0.9888876860242698

### Gradient Boosting

In [42]:
from sklearn.ensemble import GradientBoostingClassifier

In [43]:
# from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import GridSearchCV
classifiercv=GridSearchCV(classifier,param_grid=parameter,cv=5,scoring='accuracy')

In [44]:
# param_grid = {"n_estimators":[1,5,10,20,40,100],'max_depth':[3,4,5,6]}
classifiercv.fit(X_train,y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 6, 9, 12, 15, 20],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'splitter': ['best', 'random']},
             scoring='accuracy')

In [45]:
# gb_model = GradientBoostingClassifier

In [46]:
# grid = GridSearchCV(gb_model,param_grid)

In [47]:
# grid.fit(X_train,y_train)

In [48]:
from sklearn.metrics import classification_report,plot_confusion_matrix,accuracy_score

In [49]:
predication = classifiercv.predict(X_test)

In [50]:
predication

array([1, 3, 1, ..., 5, 4, 6], dtype=int64)

In [51]:
print(classification_report(y_test,predication))

              precision    recall  f1-score   support

           1       0.92      0.91      0.92       358
           2       0.89      0.92      0.90       308
           3       0.91      0.90      0.90       280
           4       0.89      0.88      0.89       412
           5       0.90      0.89      0.89       427
           6       0.98      0.99      0.99       421

    accuracy                           0.92      2206
   macro avg       0.92      0.92      0.92      2206
weighted avg       0.92      0.92      0.92      2206



### LSTM

In [52]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

In [53]:
train_df = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\X_train.txt", delim_whitespace = True,names=list(range(0,561)))

train_df['Subject'] = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\subject_train.txt", header = None, squeeze = True)

train_df["activity"] = pd.read_csv(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\y_train.txt", header = None, squeeze = True)
# C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset\train

In [54]:
features = list()
with open(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset/features.txt") as f:
    for line in f:
        features.append(line.split()[1])

In [55]:
features = list()
with open(r"C:\Users\Mitul\Downloads\UCI HAR Dataset\UCI HAR Dataset/features.txt") as f:
    for line in f:
        features.append(line.split()[1])

In [56]:
train_df.to_csv('train.csv',index_label=False)

In [57]:
test_df.to_csv('test.csv',index_label=False)

In [58]:
train_df= pd.read_csv('train.csv')

In [59]:
train_df.head()

0         1         2         3         4         5         6  \
0  0.288585 -0.020294 -0.132905 -0.995279 -0.983111 -0.913526 -0.995112   
1  0.278419 -0.016411 -0.123520 -0.998245 -0.975300 -0.960322 -0.998807   
2  0.279653 -0.019467 -0.113462 -0.995380 -0.967187 -0.978944 -0.996520   
3  0.279174 -0.026201 -0.123283 -0.996091 -0.983403 -0.990675 -0.997099   
4  0.276629 -0.016570 -0.115362 -0.998139 -0.980817 -0.990482 -0.998321   

          7         8         9  ...       553       554       555       556  \
0 -0.983185 -0.923527 -0.934724  ... -0.710304 -0.112754  0.030400 -0.464761   
1 -0.974914 -0.957686 -0.943068  ... -0.861499  0.053477 -0.007435 -0.732626   
2 -0.963668 -0.977469 -0.938692  ... -0.760104 -0.118559  0.177899  0.100699   
3 -0.982750 -0.989302 -0.938692  ... -0.482845 -0.036788 -0.012892  0.640011   
4 -0.979672 -0.990441 -0.942469  ... -0.699205  0.123320  0.122542  0.693578   

        557       558       559       560  Subject  activity  
0 -0.018446 -0.841247  0.179941 -0.058627        1         5  
1  0.703511 -0.844788  0.180289 -0.054317        1         5  
2  0.808529 -0.848933  0.180637 -0.049118        1         5  
3 -0.485366 -0.848649  0.181935 -0.047663        1         5  
4 -0.615971 -0.847865  0.185151 -0.043892        1         5  

[5 rows x 563 columns]

In [60]:
test_df= pd.read_csv('test.csv')

In [61]:
test_df.head()

0         1         2         3         4         5         6  \
0  0.257178 -0.023285 -0.014654 -0.938404 -0.920091 -0.667683 -0.952501   
1  0.286027 -0.013163 -0.119083 -0.975415 -0.967458 -0.944958 -0.986799   
2  0.275485 -0.026050 -0.118152 -0.993819 -0.969926 -0.962748 -0.994403   
3  0.270298 -0.032614 -0.117520 -0.994743 -0.973268 -0.967091 -0.995274   
4  0.274833 -0.027848 -0.129527 -0.993852 -0.967445 -0.978295 -0.994111   

          7         8         9  ...       553       554       555       556  \
0 -0.925249 -0.674302 -0.894088  ... -0.705974  0.006462  0.162920 -0.825886   
1 -0.968401 -0.945823 -0.894088  ... -0.594944 -0.083495  0.017500 -0.434375   
2 -0.970735 -0.963483 -0.939260  ... -0.640736 -0.034956  0.202302  0.064103   
3 -0.974471 -0.968897 -0.938610  ... -0.736124 -0.017067  0.154438  0.340134   
4 -0.965953 -0.977346 -0.938610  ... -0.846595 -0.002223 -0.040046  0.736715   

        557       558       559       560  Subject  activity  
0  0.271151 -0.720009  0.276801 -0.057978        2         5  
1  0.920593 -0.698091  0.281343 -0.083898        2         5  
2  0.145068 -0.702771  0.280083 -0.079346        2         5  
3  0.296407 -0.698954  0.284114 -0.077108        2         5  
4 -0.118545 -0.692245  0.290722 -0.073857        2         5  

[5 rows x 563 columns]

In [62]:
from sklearn.model_selection import train_test_split
x=train_df.drop('activity',axis=1)
y=train_df['activity']
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size= 0.30)

In [63]:
import tensorflow as tf

In [64]:
# univariate lstm example
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Flatten

In [65]:
from keras import backend as K
from keras.models import Sequential
from keras.layers import LSTM
from keras.layers.core import Dense, Dropout
from keras.layers import BatchNormalization
from keras.regularizers import L1L2

In [66]:
# Activities are the class labels
# It is a 6 class classification
ACTIVITIES = {
    0: 'WALKING',
    1: 'WALKING_UPSTAIRS',
    2: 'WALKING_DOWNSTAIRS',
    3: 'SITTING',
    4: 'STANDING',
    5: 'LAYING',
}

In [67]:
import matplotlib.pyplot as plt
import seaborn as sns

# function to print the confusion matrix

def confusion_matrix(Y_true, Y_pred):
    
    Y_true = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_true, axis=1)])
    Y_pred = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_pred, axis=1)])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

  
    plt.show() 

In [68]:
SIGNALS = ["body_acc_x","body_acc_y","body_acc_z","body_gyro_x","body_gyro_y","body_gyro_z","total_acc_x","total_acc_y","total_acc_z"]

In [69]:
# function to read the data from csv file
def _read_csv(filename):
    return pd.read_csv(filename, delim_whitespace=True, header=None)

# function to load the load
def load_signals(subset):
    signals_data = []

    for signal in SIGNALS:
        filename = f'C:/Users/Mitul/Downloads/UCI HAR Dataset/UCI HAR Dataset/{subset}/Inertial Signals/{signal}_{subset}.txt'
        
        signals_data.append(
            _read_csv(filename).to_numpy()
        ) 

   
    return np.transpose(signals_data, (1, 2, 0))




In [70]:
def load_y(subset):
    filename = f'C:/Users/Mitul/Downloads/UCI HAR Dataset/UCI HAR Dataset/{subset}/y_{subset}.txt'
    y = _read_csv(filename)[0]

    return pd.get_dummies(y).to_numpy()

In [71]:
def load_data():
    X_train, X_test = load_signals('train'), load_signals('test')
    y_train, y_test = load_y('train'), load_y('test')

    return X_train, X_test, y_train, y_test

In [72]:
# Importing tensorflow
np.random.seed(42)
import tensorflow as tf
tf.random.set_seed(42)

In [73]:
# Initializing parameters
epochs = 50
batch_size = 16
n_hidden = 32

In [74]:
#function to count the number of classes
def _count_classes(y):
    return len(set([tuple(category) for category in y]))

In [75]:
X_train, X_test, Y_train, Y_test = load_data()

In [76]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = _count_classes(Y_train)

print(timesteps)
print(input_dim)
print(len(X_train))

128
9
7352


In [77]:
# Initiliazing the sequential model
model = Sequential()
# Configuring the parameters
model.add(LSTM(n_hidden, input_shape=(timesteps, input_dim)))
# Adding a dropout layer
model.add(Dropout(0.5))
# Adding a dense output layer with sigmoid activation
model.add(Dense(n_classes, activation='sigmoid'))
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm (LSTM)                 (None, 32)                5376      
                                                                 
 dropout (Dropout)           (None, 32)                0         
                                                                 
 dense (Dense)               (None, 6)                 198       
                                                                 
Total params: 5,574
Trainable params: 5,574
Non-trainable params: 0
_________________________________________________________________


In [78]:
# Compiling the model
model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

In [79]:
# Training the model
model.fit(X_train, Y_train, batch_size=batch_size, validation_data=(X_test, Y_test),epochs=epochs)

Epoch 1/50
460/460 [==============================] - 27s 42ms/step - loss: 1.2041 - accuracy: 0.4910 - val_loss: 1.0376 - val_accuracy: 0.5826
Epoch 2/50
460/460 [==============================] - 16s 34ms/step - loss: 0.7777 - accuracy: 0.6723 - val_loss: 0.7371 - val_accuracy: 0.7167
Epoch 3/50
460/460 [==============================] - 15s 33ms/step - loss: 0.6163 - accuracy: 0.7414 - val_loss: 0.6360 - val_accuracy: 0.7727
Epoch 4/50
460/460 [==============================] - 16s 35ms/step - loss: 0.4841 - accuracy: 0.8213 - val_loss: 0.5668 - val_accuracy: 0.8215
Epoch 5/50
460/460 [==============================] - 15s 34ms/step - loss: 0.3613 - accuracy: 0.8882 - val_loss: 0.4281 - val_accuracy: 0.8602
Epoch 6/50
460/460 [==============================] - 16s 34ms/step - loss: 0.2793 - accuracy: 0.9154 - val_loss: 0.4235 - val_accuracy: 0.8636
Epoch 7/50
460/460 [==============================] - 16s 34ms/step - loss: 0.2483 - accuracy: 0.9236 - val_loss: 0.3636 - val_accuracy:

In [80]:
(Y_test, model.predict(X_test))

93/93 [==============================] - 1s 9ms/step


(array([[0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        ...,
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0]], dtype=uint8),
 array([[6.2890700e-05, 4.7086683e-05, 3.4460729e-05, 7.4061149e-01,
         9.9380958e-01, 6.2238185e-05],
        [5.8878992e-05, 4.5231198e-05, 3.0798325e-05, 7.4832189e-01,
         9.9388057e-01, 5.8050406e-05],
        [5.6426001e-05, 4.2954449e-05, 3.0141186e-05, 7.4419564e-01,
         9.9424624e-01, 5.3682143e-05],
        ...,
        [4.2678651e-01, 9.9999934e-01, 1.9161302e-01, 2.1208435e-01,
         1.4665303e-02, 8.4658220e-02],
        [3.2646957e-01, 9.9999827e-01, 9.1089487e-02, 2.6400742e-01,
         2.1383060e-02, 9.0627141e-02],
        [3.7136051e-01, 9.9999797e-01, 7.7679127e-02, 2.4255905e-01,
         4.8822656e-02, 1.9417889e-01]], dtype=float32))

In [81]:
score = model.evaluate(X_test, Y_test)

print("\n   cat_crossentropy  ||   accuracy ")
print("  ____________________________________")
print(score)

93/93 [==============================] - 1s 9ms/step - loss: 0.5662 - accuracy: 0.9094

   cat_crossentropy  ||   accuracy 
  ____________________________________
[0.5662268400192261, 0.9093993902206421]


In [82]:
# Initializing parameters
n_epochs = 50
n_batch = 16
n_classes = _count_classes(Y_train)

# Bias regularizer value - we will use elasticnet
reg = L1L2(0.01, 0.01)

In [83]:
# Model execution
model = Sequential()
model.add(LSTM(48, input_shape=(timesteps, input_dim), return_sequences=True,bias_regularizer=reg ))
model.add(BatchNormalization())
model.add(Dropout(0.50))
model.add(LSTM(32))
model.add(Dropout(0.50))
model.add(Dense(n_classes, activation='sigmoid'))
print("Model Summary: ")
model.summary()

Model Summary: 
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm_1 (LSTM)               (None, 128, 48)           11136     
                                                                 
 batch_normalization (BatchN  (None, 128, 48)          192       
 ormalization)                                                   
                                                                 
 dropout_1 (Dropout)         (None, 128, 48)           0         
                                                                 
 lstm_2 (LSTM)               (None, 32)                10368     
                                                                 
 dropout_2 (Dropout)         (None, 32)                0         
                                                                 
 dense_1 (Dense)             (None, 6)                 198       
                                      

In [84]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [85]:
# Training the model
model.fit(X_train, Y_train, batch_size=n_batch, validation_data=(X_test, Y_test), epochs=n_epochs)

Epoch 1/50
460/460 [==============================] - 34s 64ms/step - loss: 1.5695 - accuracy: 0.6612 - val_loss: 1.6585 - val_accuracy: 0.6542
Epoch 2/50
460/460 [==============================] - 2930s 6s/step - loss: 0.7843 - accuracy: 0.8493 - val_loss: 0.6101 - val_accuracy: 0.8473
Epoch 3/50
460/460 [==============================] - 49s 106ms/step - loss: 0.4330 - accuracy: 0.9011 - val_loss: 0.3034 - val_accuracy: 0.9152
Epoch 4/50
460/460 [==============================] - 38s 83ms/step - loss: 0.2804 - accuracy: 0.9230 - val_loss: 0.3156 - val_accuracy: 0.8890
Epoch 5/50
460/460 [==============================] - 37s 80ms/step - loss: 0.2184 - accuracy: 0.9302 - val_loss: 0.2924 - val_accuracy: 0.8965
Epoch 6/50
460/460 [==============================] - 33s 71ms/step - loss: 0.2079 - accuracy: 0.9290 - val_loss: 0.2819 - val_accuracy: 0.9087
Epoch 7/50
460/460 [==============================] - 32s 69ms/step - loss: 0.1673 - accuracy: 0.9416 - val_loss: 0.2455 - val_accuracy

In [86]:
# Confusion Matrix
#new_confusion_matrix(Y_test, model.predict(X_test))
Y_test, model.predict(X_test)

93/93 [==============================] - 4s 25ms/step


(array([[0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        ...,
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0]], dtype=uint8),
 array([[4.78576217e-03, 3.17483884e-03, 6.49758265e-04, 9.66304421e-01,
         9.98616099e-01, 1.01245870e-03],
        [4.43624100e-03, 2.74694175e-03, 5.49788587e-04, 9.71669316e-01,
         9.98772204e-01, 8.40341905e-04],
        [4.39720042e-03, 2.49139545e-03, 5.21706359e-04, 9.73361015e-01,
         9.98804688e-01, 8.48239521e-04],
        ...,
        [3.24507833e-01, 9.99995828e-01, 4.64380056e-01, 1.27935968e-02,
         5.98467365e-02, 1.44668221e-02],
        [3.85465413e-01, 9.99997437e-01, 4.82960790e-01, 1.36658698e-02,
         7.96027780e-02, 1.51814334e-02],
        [3.72803509e-01, 9.99998331e-01, 4.15793568e-01, 1.22035164e-02,
         1.06871076e-01, 1.87899768e-02]], dtype=float32))

In [87]:
score = model.evaluate(X_test, Y_test)

print("\n   cat_crossentropy  ||   accuracy ")
print("  ____________________________________")
print(score)

93/93 [==============================] - 2s 24ms/step - loss: 0.3460 - accuracy: 0.9264

   cat_crossentropy  ||   accuracy 
  ____________________________________
[0.346041202545166, 0.9263657927513123]


### Accuracy

#### Logistic Regression:- 98.38%

#### Decision Tree:-95.08%

#### Gradient Boosting:-91%

#### LSTM:-92.19%